plugins/OnlineSales.Plugin.SendGrid/Controllers/SendgridController.cs

```diff
@@ -0,0 +1,275 @@
+{
+    private static readonly int BatchSize = 500;    
+
+    private readonly SendgridDbContext dbContext;
+
+    private readonly IContactService contactService;
+
+    public SendgridController(SendgridDbContext dbContext, EsDbContext esDbContext, IConfiguration configuration, IContactService contactService)
+    {
+        this.dbContext = dbContext;
+        this.contactService = contactService;
+    }
+
+    [HttpPost]
+    [Authorize]
+    [Route("import")]
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status500InternalServerError)]
+    public async Task<ActionResult> Import([FromBody] List<MessageEventDto> records)
+    {
+        return await AddEvents(records);
+    }
+
+    [HttpPost]
+    [AllowAnonymous]
+    [Route("webhook")]
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status500InternalServerError)]
+    public async Task<ActionResult> AddMessageEvents()
+    {
+        var signature = Request.Headers[RequestValidator.SIGNATURE_HEADER].First();
+        var timestamp = Request.Headers[RequestValidator.TIMESTAMP_HEADER].First();
+        var body = await GetRequestBody(HttpContext.Request.Body);
+
+        if (signature == null || timestamp == null || !Verify(body, signature, timestamp))
+        {
+            throw new SendGridApiException("Webhook event can't be verified");
+        }
+
+        var records = JsonConvert.DeserializeObject<List<WebhookEventDto>>(body);
+
+        if (records == null)
+        {
+            throw new SendGridApiException("Webhook event body can't be deserialized");
+        }
+
+        return await AddEvents(records);
+    }
+
+    private async Task<Dictionary<Contact, List<T>>> CreateNonExistedContacts<T>(IEnumerable<IGrouping<string, T>> emailAndRecords)
+    {
+        var contactRecords = new Dictionary<Contact, List<T>>();
+        var nonExistedContacts = new List<Contact>();
+
+        int position = 0;
+
+        while (position < emailAndRecords.Count())
+        {
+            var batch = emailAndRecords.Skip(position).Take(BatchSize);
+            var existedContacts = dbContext.Contacts!.Where(c => batch.Select(b => b.Key).Contains(c.Email)).ToDictionary(c => c.Email, c => c);
+            foreach (var b in batch)
+            {
+                var isExists = existedContacts.TryGetValue(b.Key, out var contact);
+                if (!isExists)
+                {
+                    contact = new Contact() { Email = b.Key };
+                    nonExistedContacts.Add(contact);
+                }
+
+                contactRecords[contact!] = b.ToList();
+            }
+
+            position += BatchSize;
+        }
+
+        await contactService.EnrichWithDomainIdAsync(nonExistedContacts);
```

Comment:
@IgorKostyukov EnrichWithDomainIdAsync method should not be used directly from the outside of the service it-selves as that logic is internal for the service and may be changed. The service should expose methods to save single contact or a list of contacts and all the Enriching logic should happen inside of it rather than in different places outside the service, especially if it is a plugin. If you use it from outside the logic and as the result the data in DB may become inconsistent over. It is a very good sample here as well, in the service we also have logic to enrich contacts with AccountId, but it is not considered here. Please refactor.